File: apps/api/app/schemas/market.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel


class ExchangeResponse(BaseModel):
    code: str
    name: str
    timezone: str


class CompanyResponse(BaseModel):
    id: str
    symbol: str
    name: str
    sector: str
    exchange: ExchangeResponse
    official_website: str | None = None
    psx_url: str | None = None
    description: str | None = None
    is_active: bool


class MarketPriceResponse(BaseModel):
    symbol: str
    trade_date: date
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    previous_close: Decimal
    change: Decimal
    change_percent: Decimal
    volume: int
    value: Decimal
    market_cap: Decimal | None = None
    source: str
    source_url: str | None = None
    ingested_at: datetime | None


class MarketSnapshotResponse(BaseModel):
    snapshot_date: date
    index_name: str
    index_value: Decimal
    index_change: Decimal
    index_change_percent: Decimal
    total_volume: int
    total_value: Decimal
    source: str
    ingested_at: datetime


class SectorDailyStatsResponse(BaseModel):
    sector: str
    trade_date: date
    total_volume: int
    total_value: Decimal
    average_change_percent: Decimal
    advancers: int
    decliners: int
    unchanged: int
    source: str


class CompanyDetailResponse(BaseModel):
    company: CompanyResponse
    latest_price: MarketPriceResponse | None = None


class MarketOverviewResponse(BaseModel):
    snapshot: MarketSnapshotResponse | None
    top_gainers: list[MarketPriceResponse]
    top_losers: list[MarketPriceResponse]
    top_volume: list[MarketPriceResponse]
    sectors: list[SectorDailyStatsResponse]


class MarketFreshnessResponse(BaseModel):
    market_data_mode: str
    refresh_seconds: int
    last_successful_ingestion_at: datetime | None
    latest_trade_date: date | None
    latest_source: str | None
    latest_attempted_provider: str | None = None
    latest_used_provider: str | None = None
    is_stale: bool
    stale_warning: str | None
    backup_warning: str | None = None
    # Split fields: "stale" previously conflated mock-mode, ingestion age, and
    # trade-date/session status into one boolean and one warning string.
    ingestion_age_seconds: float | None = None
    provider_mode_warning: str | None = None
    ingestion_staleness_warning: str | None = None
    fallback_provider_active: bool = False
    trade_date_status: Literal["current", "prior_session", "stale", "unknown"] = "unknown"
    exchange_session_status: Literal["open", "closed", "unknown"] = "unknown"
    exchange_session_note: str | None = None

File: apps/web/lib/api/market.ts
import { request } from "@/lib/api/client";

export type Exchange={code:string;name:string;timezone:string};
export type Company={id:string;symbol:string;name:string;sector:string;exchange:Exchange;official_website?:string|null;psx_url?:string|null;description?:string|null;is_active:boolean};
export type MarketPrice={symbol:string;trade_date:string;open:string;high:string;low:string;close:string;previous_close:string;change:string;change_percent:string;volume:number;value:string;market_cap?:string|null;source:string;source_url?:string|null;ingested_at:string};
export type MarketSnapshot={snapshot_date:string;index_name:string;index_value:string;index_change:string;index_change_percent:string;total_volume:number;total_value:string;source:string;ingested_at:string};
export type SectorDailyStats={sector:string;trade_date:string;total_volume:number;total_value:string;average_change_percent:string;advancers:number;decliners:number;unchanged:number;source:string};
export type CompanyDetail={company:Company;latest_price?:MarketPrice|null};
export type MarketOverview={snapshot?:MarketSnapshot|null;top_gainers:MarketPrice[];top_losers:MarketPrice[];top_volume:MarketPrice[];sectors:SectorDailyStats[]};
export type MarketFreshness={market_data_mode:string;refresh_seconds:number;last_successful_ingestion_at?:string|null;latest_trade_date?:string|null;latest_source?:string|null;latest_attempted_provider?:string|null;latest_used_provider?:string|null;is_stale:boolean;stale_warning?:string|null;backup_warning?:string|null;ingestion_age_seconds?:number|null;provider_mode_warning?:string|null;ingestion_staleness_warning?:string|null;fallback_provider_active:boolean;trade_date_status:"current"|"prior_session"|"stale"|"unknown";exchange_session_status:"open"|"closed"|"unknown";exchange_session_note?:string|null};
export type SourceHealth={source:string;status:"healthy"|"stale"|"partial"|"failed"|"never_run";last_attempt?:string|null;last_success?:string|null;latest_data_at?:string|null;attempted:number;accepted:number;updated:number;rejected:number;error?:string|null;freshness_sla_minutes?:number|null};
export type DataHealth={sources:SourceHealth[]};
export type CoverageCategory={available:boolean;count:number;latest_date?:string|null;reason?:string|null};
export type CompanyCompleteness={symbol:string;price:CoverageCategory&{observations:number};fundamentals:CoverageCategory&{fact_count:number;latest_period?:string|null};reports:CoverageCategory;announcements:CoverageCategory;news:CoverageCategory};

export const getMarketOverview=()=>request<MarketOverview>("/market/overview");
export const getMarketFreshness=()=>request<MarketFreshness>("/market/freshness");
export const getDataHealth=()=>request<DataHealth>("/ingestion/health");
export const getCompanyCompleteness=(symbol:string)=>request<CompanyCompleteness>(`/ingestion/companies/${encodeURIComponent(symbol)}/completeness`);
export function getCompanies(query?:string,options?:{signal?:AbortSignal}){const params=new URLSearchParams({limit:"1000"});if(query)params.set("q",query);return request<Company[]>(`/market/companies?${params.toString()}`,options?.signal?{signal:options.signal}:{})}
export const getCompanyDetail=(symbol:string)=>request<CompanyDetail>(`/market/company/${encodeURIComponent(symbol)}`);
export const getCompanyHistory=(symbol:string,limit=2000)=>request<MarketPrice[]>(`/market/company/${encodeURIComponent(symbol)}/history?limit=${encodeURIComponent(String(limit))}`);
